django/core/signing.py

```diff
@@ -269,12 +277,28 @@ def unsign(self, value, max_age=None):
         """
         result = super().unsign(value)
         value, timestamp = result.rsplit(self.sep, 1)
-        timestamp = b62_decode(timestamp)
-        if max_age is not None:
-            if isinstance(max_age, datetime.timedelta):
-                max_age = max_age.total_seconds()
-            # Check timestamp is not older than max_age
-            age = time.time() - timestamp
-            if age > max_age:
-                raise SignatureExpired("Signature age %s > %s seconds" % (age, max_age))
+        self.payload_timestamp = b62_decode(timestamp)
+        self._verify_max_age(max_age)
+        return value
+
+    def unsign_object(self, signed_obj, max_age=None, **kwargs):
+        if self.expiration_key is None:
+            return super(TimestampSigner, self).unsign_object(
+                signed_obj, max_age=max_age, **kwargs
+            )
+
+        # If using expiration key use it's value when present
+        value = super().unsign_object(signed_obj, **kwargs, max_age=None)
+        max_age = value.get(self.expiration_key, max_age)
+        self._verify_max_age(max_age)
         return value
+
+    def _verify_max_age(self, max_age=None):
+        if max_age is None:
+            return
+        if isinstance(max_age, datetime.timedelta):
+            max_age = max_age.total_seconds()
+        # Check timestamp is not older than max_age
+        age = time.time() - self.payload_timestamp
+        if age > max_age:
+            raise SignatureExpired("Signature age %s > %s seconds" % (age, max_age))
```

Comment:
We could add this hook in a separate commit (1st) and move the main change to the 2nd commit.